### PR TITLE
DefaultCommonIdeaService code refactoring, migration of common logic to HybrisProjectSettingsComponent

### PR DESCRIPTION
**Pull request by absinthminded**

### CODE REVIEW

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 1]
```diff
@@ -49,33 +44,34 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
         )
+
         return matchAllModuleNames(webservicesNames, moduleNames)
     }
 
+    private fun matchAllModuleNames(
+        namePatterns: Collection<String>,
+        moduleNames: Collection<String>
+    ): Boolean {
+        return namePatterns.all { pattern: String -> matchModuleName(pattern, moduleNames) }
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: optimize:
- return right away
- no need in `pattern` variable

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 2]
```diff
@@ -89,87 +85,62 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
-        var mySettings = settings
-        if (mySettings == null) {
-            mySettings = HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
-                .getActiveHacRemoteConnectionSettings(project)
-        }
-        return mySettings?.sslProtocol ?: HybrisConstants.DEFAULT_SSL_PROTOCOL
+        val currentSettings = settings ?: HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
+            .getActiveHacRemoteConnectionSettings(project)
+
+        return currentSettings?.sslProtocol ?: HybrisConstants.DEFAULT_SSL_PROTOCOL
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: return currentSettings
?.sslProtocol 
?: HybrisConstants.DEFAULT_SSL_PROTOCOL

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 3]
```diff
@@ -89,87 +85,62 @@ class DefaultCommonIdeaService : CommonIdeaService {
     }
 
     override fun getActiveSslProtocol(project: Project, settings: HybrisRemoteConnectionSettings?): String {
-        var mySettings = settings
-        if (mySettings == null) {
-            mySettings = HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
-                .getActiveHacRemoteConnectionSettings(project)
-        }
-        return mySettings?.sslProtocol ?: HybrisConstants.DEFAULT_SSL_PROTOCOL
+        val currentSettings = settings ?: HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: extract to separate method and reuse also in `getHostHacUrl`, once done, review optimization of the methods (code reduction)

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 4]
```diff
@@ -89,87 +85,62 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
-        return getUrl(mySettings)
+        val currentSettings = settings ?: HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
+            .getActiveHacRemoteConnectionSettings(project)
+
+        return getUrl(currentSettings)
     }
 
     override fun getSolrUrl(project: Project, settings: HybrisRemoteConnectionSettings?): String {
-        var mySettings = settings
+        var currentSettings = settings
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: reuse same pattern for getting fallback settings

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 5]
```diff
@@ -179,37 +150,11 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
-        return namePatterns.stream()
-            .allMatch { pattern: String -> matchModuleName(pattern, moduleNames) }
-    }
-
     private fun matchModuleName(pattern: String, moduleNames: Collection<String>): Boolean {
         val regex = Regex("\\Q$pattern\\E".replace("*", "\\E.*\\Q"))
         return moduleNames.stream()
             .parallel()
-            .anyMatch { p: String -> p.matches(regex) }
+            .anyMatch { it: String -> it.matches(regex) }
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: no need in `it: String`, Intellij IDEA had to suggest that

File: src/com/intellij/idea/plugin/hybris/project/descriptors/RegularHybrisModuleDescriptor.java [hunk 6]
```diff
@@ -252,29 +231,38 @@ public List<JavaLibraryDescriptor> getLibraryDescriptors() {
                 });
         }
 
+        final Project currentProject = getRootProjectDescriptor().getProject();
+        if (currentProject != null) {
+            addLibsForBackofficeModule(libs, currentProject);
+        }
+
+        if (this.isAddOn()) {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/project/descriptors/RegularHybrisModuleDescriptor.java:239]: not sure that `this` is required here, consider removing it

File: src/com/intellij/idea/plugin/hybris/project/descriptors/RegularHybrisModuleDescriptor.java [hunk 7]
```diff
@@ -252,29 +231,38 @@ public List<JavaLibraryDescriptor> getLibraryDescriptors() {
...
+    private void addLibsForBackofficeModule(final List<JavaLibraryDescriptor> libs, final Project currentProject) {
         if (this.hasBackofficeModule()) {
 
+            final HybrisProjectSettingsComponent instance = HybrisProjectSettingsComponent.getInstance(currentProject);
+
             libs.add(new DefaultJavaLibraryDescriptor(
                 new File(
                     this.getRootProjectDescriptor().getHybrisDistributionDirectory(),
-                    CommonIdeaService.Companion.getInstance().getBackofficeWebInfLib(getRootProjectDescriptor().getProject())
+                    instance.getBackofficeWebInfLib()
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/project/descriptors/RegularHybrisModuleDescriptor.java]: do not use static access to external fields with field names which may cause collisions.

`instance` -> `MyService.instance`

File: src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt [hunk 8]
```diff
@@ -37,6 +41,14 @@ class HybrisProjectSettingsComponent : PersistentStateComponent<HybrisProjectSet
 
     fun isHybrisProject() = state.hybrisProject
 
+    fun isOutdatedHybrisProject(): Boolean {
+        val lastImportVersion = hybrisProjectSettings.importedByVersion ?: return true
+        val currentVersion = PluginManagerCore.getPlugin(PluginId.getId(HybrisConstants.PLUGIN_ID))?.version
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt]:
```
...ts.PLUGIN_ID))
?.ver....
```

File: src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt [hunk 9]
```diff
@@ -63,6 +75,22 @@ class HybrisProjectSettingsComponent : PersistentStateComponent<HybrisProjectSet
     fun registerCloudExtensions() = HybrisConstants.CCV2_COMMERCE_CLOUD_EXTENSIONS
             .forEach { state.availableExtensions[it] = ExtensionDescriptor(it, HybrisModuleDescriptorType.CCV2) }
 
+    fun getBackofficeWebInfLib(): String {
+        return if (is2019VersionOrHigher()) HybrisConstants.BACKOFFICE_WEB_INF_LIB_2019 else HybrisConstants.BACKOFFICE_WEB_INF_LIB
+    }
+
+    fun getBackofficeWebInfClasses(): String {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt]: inline

File: src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt [hunk 10]
```diff
@@ -63,6 +75,22 @@ class HybrisProjectSettingsComponent : PersistentStateComponent<HybrisProjectSet
     fun registerCloudExtensions() = HybrisConstants.CCV2_COMMERCE_CLOUD_EXTENSIONS
             .forEach { state.availableExtensions[it] = ExtensionDescriptor(it, HybrisModuleDescriptorType.CCV2) }
 
+    fun getBackofficeWebInfLib(): String {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt]: inline

File: src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt [hunk 11]
```diff
@@ -63,6 +75,22 @@ class HybrisProjectSettingsComponent : PersistentStateComponent<HybrisProjectSet
...
+    fun getBackofficeWebInfClasses(): String {
+        return if (is2019VersionOrHigher()) HybrisConstants.BACKOFFICE_WEB_INF_CLASSES_2019 else HybrisConstants.BACKOFFICE_WEB_INF_CLASSES
+    }
+
+    private fun is2019VersionOrHigher(): Boolean {
+        val hybrisVersion = state.hybrisVersion
+        if (hybrisVersion.isNullOrBlank()) return false
+
+        val projectVersion = Version.parseVersion(hybrisVersion)
+        return projectVersion >= Version.parseVersion("1905.0")
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/settings/HybrisProjectSettingsComponent.kt]: feel free to introduce new constant -> PLATFORM_VERSION_1905_0 = "1905.0"

File: src/com/intellij/idea/plugin/hybris/startup/HybrisProjectStructureStartupActivity.kt [hunk 12]
```diff
@@ -51,18 +51,19 @@ class HybrisProjectStructureStartupActivity : ProjectActivity {
         if (project.isDisposed) return
 
         val commonIdeaService = ApplicationManager.getApplication().getService(CommonIdeaService::class.java)
-        val isHybrisProject = HybrisProjectSettingsComponent.getInstance(project).isHybrisProject()
+        val instance = HybrisProjectSettingsComponent.getInstance(project)
+        val isHybrisProject = instance.isHybrisProject()
 
         if (isHybrisProject) {
-            commonIdeaService.refreshProjectSettings(project)
+            instance.registerCloudExtensions()
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/startup/HybrisProjectStructureStartupActivity.kt]: mby not `instance`, rather `projectSettings`

File: src/com/intellij/idea/plugin/hybris/project/descriptors/RegularHybrisModuleDescriptor.java (hunk 6 above)
Comment (absinthminded) [reply src/com/intellij/idea/plugin/hybris/project/descriptors/RegularHybrisModuleDescriptor.java:239]: strange git representation. anyway let it be like this for now and wait it's queue

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 13]
```diff
@@ -89,87 +80,59 @@ class DefaultCommonIdeaService : CommonIdeaService {
     }
 
     override fun getActiveSslProtocol(project: Project, settings: HybrisRemoteConnectionSettings?): String {
-        var mySettings = settings
-        if (mySettings == null) {
-            mySettings = HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
-                .getActiveHacRemoteConnectionSettings(project)
-        }
-        return mySettings?.sslProtocol ?: HybrisConstants.DEFAULT_SSL_PROTOCOL
+        val currentSettings = getProjectSettings(project, settings)
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: no need in variable, function can return statement right away

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 14]
```diff
@@ -89,87 +80,59 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
     }
 
     override fun getHostHacUrl(project: Project, settings: HybrisRemoteConnectionSettings?): String {
-        var mySettings = settings
-        if (mySettings == null) {
-            mySettings = HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
-                .getActiveHacRemoteConnectionSettings(project)
-        }
-        return getUrl(mySettings)
+        val currentSettings = getProjectSettings(project, settings)
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: no need in variable, function can return statement right away

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 15]
```diff
@@ -89,87 +80,59 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
-        sb.append(mySettings.hostIP)
-        sb.append(":")
-        sb.append(mySettings.port)
-        sb.append("/")
-        sb.append(mySettings.solrWebroot)
-        val result = sb.toString()
-        LOG.debug("Calculated host SOLR URL=$result")
-        return result
+    private fun getProjectSettings(project: Project, settings: HybrisRemoteConnectionSettings?): HybrisRemoteConnectionSettings? {
+        return settings ?: HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: function can return statement right away

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 16]
```diff
@@ -89,87 +80,59 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
-        return projectVersion.compareTo(_1905) >= 0
-    }
+    override fun getSolrUrl(project: Project, settings: HybrisRemoteConnectionSettings?): String {
+        val currentSettings = settings ?: HybrisDeveloperSpecificProjectSettingsComponent.getInstance(project)
+            .getActiveSolrRemoteConnectionSettings(project)
 
-    override fun getBackofficeWebInfLib(project: Project): String {
-        return if (is2019plus(project)) HybrisConstants.BACKOFFICE_WEB_INF_LIB_2019 else HybrisConstants.BACKOFFICE_WEB_INF_LIB
-    }
+        val solrUrl = StringBuilder()
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: usage of the StringBuilder is not required here, please, refactor the code

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 17]
```diff
@@ -179,62 +142,29 @@ class DefaultCommonIdeaService : CommonIdeaService {
...
-        val hac = settings.hacWebroot
-        if (StringUtils.isNoneBlank(hac)) {
-            sb.append('/')
-            sb.append(StringUtils.strip(hac, " /"))
-        }
-        val result = sb.toString()
-        LOG.debug("Calculated hostHacURL=$result")
-        return result
-    }
+        val urlBuilder = StringBuilder()
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: here, StringBuilder can be used, but same can be also achieved with Kotlin StringTemplates, slightly different approach

File: src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt [hunk 18]
```diff
@@ -17,224 +17,141 @@
...
     override fun isTypingActionInProgress(): Boolean {
-        val isTyping = StringUtils.equalsAnyIgnoreCase(
-            commandProcessor.currentCommandName, *HybrisConstants.TYPING_EDITOR_ACTIONS
-        )
-        val isUndoOrRedo = StringUtils.startsWithAny(
-            commandProcessor.currentCommandName, *HybrisConstants.UNDO_REDO_EDITOR_ACTIONS
-        )
-        return isTyping || isUndoOrRedo
-    }
+        val isTyping = HybrisConstants.TYPING_EDITOR_ACTIONS.any {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/common/services/impl/DefaultCommonIdeaService.kt]: extract `commandProcessor.currentCommandName` to variable, why do the same multimple time

**Review by mlytvyn — APPROVED** [review]

**Pull request merged**